cli2gui/application/application.py

```diff
@@ -381,33 +561,547 @@ def run(buildSpec: c2gtypes.FullBuildSpec):
+	window['-terminal-'].bind("<BackSpace>", "+BackSpace+", propagate=False) # Element.bind(propagate=x) requires https://github.com/PySimpleGUI/PySimpleGUI/pull/5315
+	window['-terminal-'].bind("<Tab>", "+Tab+", propagate=False) # Element.bind(propagate=x) requires https://github.com/PySimpleGUI/PySimpleGUI/pull/5315
+	window['-terminal-'].bind("<Command-k>", "+Command-k+")
+	window['-terminal-'].bind("<Command-c>", "+Command-c+")
+
+	window['-terminal-'].bind("<Left>", "+Left+", propagate=False)
+	window['-terminal-'].bind("<Right>", "+Right+", propagate=False)
+	window['-terminal-'].bind("<Up>", "+Up+", propagate=False)
+	window['-terminal-'].bind("<Down>", "+Down+", propagate=False)
+
+	# pointer events (mouse events)
+	#window['-terminal-'].bind("<Enter>", "+Enter+") # mouse enter
+	window['-terminal-'].bind("<Button-1>", "+Button-1+", propagate=False) # ClickLeft
+	window['-terminal-'].bind("<Button-2>", "+Button-2+", propagate=False) # ClickMiddle
+	#window['-terminal-'].bind("<Button-3>", "+Button-3+") # ClickRight
+
+	#shell_process_command = 'bash'
+	#shell_process_command = ['busybox', 'sh']
+	shell_process_command = ['busybox', 'sh', '+m'] # completely disable job control = remove the initial warning "job control turned off"
+	#shell_process_command = ['busybox', 'sh', '-c', 'setsid cttyhack sh; exec sh'] # cttyhack: can't open '/dev/tty0': Permission denied
+	#shell_process_command = ['busybox', 'sh', '-c', 'setsid -c sh; exec sh'] # exec /bin/sh
+	#shell_process_command = ['busybox', 'sh', '-c', 'exec setsid -c -w sh'] # exec /bin/sh
+	# https://stackoverflow.com/questions/36529881/qemu-bin-sh-cant-access-tty-job-control-turned-off
+
+	# NOTE this is just a posix shell, no bash shell. TODO test on windows
+
+	# open pseudo-terminal to interact with subprocess
+	master_fd, slave_fd = pty.openpty()
+
+	def gui_to_shell_string(string):
+		bytes_ = string.encode("utf8")
+		os.write(master_fd, bytes_)
+
+	def gui_to_shell_bytes(bytes_):
+		os.write(master_fd, bytes_)
+
+	# unix only: use os.setsid() make it run in a new process group, or bash job control will not be enabled
+	shell_env = os.environ.copy() # this seems to fix darkmode. weird.
+	shell_process = subprocess.Popen(
+		shell_process_command,
+		preexec_fn=os.setsid,
+		stdin=slave_fd,
+		stdout=slave_fd,
+		stderr=slave_fd,
+		universal_newlines=True,
+		env=shell_env,
+		encoding="utf8",
+		close_fds=True,
+	)
+	log.debug(f"started shell process with PID {shell_process.pid}")
+
+	child_process_list = []
+	child_pid_set_before_run = None
+	child_process_list_after_run = None
+	def child_process_watcher():
+
+		# FIXME avoid global, use queue or lock https://stackoverflow.com/a/19790994/10440128
+		global child_process_list, child_pid_set_before_run, child_process_list_after_run
+
+		proc = psutil.Process(pid=shell_process.pid)
+		# https://psutil.readthedocs.io/en/latest/#process-class
+		while True:
+			log.debug(f'child_process_watcher ---------------------------')
```

Comment:
Looks like an f-string isn't needed here. pylint should scream about this all being well